feat: Add Hyperliquid trading functionality and components

sidmorizon opened the pull request:
- Integrate Hyperliquid trading features with new components and hooks for trading operations
- Update package.json and yarn.lock to add @nktkas/hyperliquid dependency
- Add trading panel, order confirmation, leverage adjustment modules for improved UX
- Implement Hyperliquid-related state management and data processing logic

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Added a Perp trading tab with Hyperliquid integration: market/limit orders, TP/SL, leverage adjustment, and account overview.
  - Introduced token selector, ticker bar, order book, and initial candles view.
  - Provided positions, open orders, and trade history (with mobile/desktop layouts and a history modal).
  - Implemented wallet status checks and streamlined trading enablement.

- Chores
  - Added Hyperliquid SDK dependency.
  - Expanded routing and modal support for Perp.
  - Extended event bus with Hyperliquid data and connection events.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Minnzen committed:
feat: Add Hyperliquid trading functionality and components

- Integrate Hyperliquid trading features with new components and hooks for trading operations
- Update package.json and yarn.lock to add @nktkas/hyperliquid dependency
- Add trading panel, order confirmation, leverage adjustment modules for improved UX
- Implement Hyperliquid-related state management and data processing logic

Minnzen committed:
feat: Add Hyperliquid trading functionality and components

- Integrate Hyperliquid trading features with new components and hooks for trading operations
- Update package.json and yarn.lock to add @nktkas/hyperliquid dependency
- Add trading panel, order confirmation, leverage adjustment modules for improved UX
- Implement Hyperliquid-related state management and data processing logic

revan-zhang commented:
### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/rayston92/pr-checks/5361cfff-e029-490d-88d0-79671c7974ff?refs=d606d6485d3496d6ecb759b83289d252173b0527&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/rayston92/pr-checks/5361cfff-e029-490d-88d0-79671c7974ff/license?refs=d606d6485d3496d6ecb759b83289d252173b0527&source=prChecksComment)

Minnzen committed:
refactor: consolidate Hyperliquid type definitions and remove redundant files

- Removed 4 redundant type files containing duplicate definitions
- Created unified business.ts for business logic types
- Simplified sdk.ts to focus on SDK type re-exports using @nktkas/hyperliquid namespace
- Streamlined index.ts as central export point with validation constants and error messages
- Updated all Hyperliquid services, components, and hooks to use new consolidated type structure

Minnzen committed:
refactor: consolidate Hyperliquid type definitions and remove redundant files

- Removed 4 redundant type files containing duplicate definitions
- Created unified business.ts for business logic types
- Simplified sdk.ts to focus on SDK type re-exports using @nktkas/hyperliquid namespace
- Streamlined index.ts as central export point with validation constants and error messages
- Updated all Hyperliquid services, components, and hooks to use new consolidated type structure

Minnzen committed:
refactor: optimize Hyperliquid state management and remove redundant utilities

- Consolidate data processing logic from DataProcessor.ts into atoms.ts for better state management
- Remove SubscriptionConfig.ts and integrate subscription logic into Jotai context
- Simplify hooks implementation by leveraging enhanced atom-based state management

Minnzen committed:
refactor: optimize Hyperliquid state management and remove redundant utilities

- Consolidate data processing logic from DataProcessor.ts into atoms.ts for better state management
- Remove SubscriptionConfig.ts and integrate subscription logic into Jotai context
- Simplify hooks implementation by leveraging enhanced atom-based state management

Minnzen committed:
feat: implement order book data management and processing

- Add l2BookAtom for order book state management in Jotai context
- Implement useL2Book hook with order book data processing utilities
- Add order book data interfaces and calculation methods for spread, volume analysis
- Include basic PerpOrderBook component as data usage example

Minnzen committed:
feat: implement order book data management and processing

- Add l2BookAtom for order book state management in Jotai context
- Implement useL2Book hook with order book data processing utilities
- Add order book data interfaces and calculation methods for spread, volume analysis
- Include basic PerpOrderBook component as data usage example

Minnzen committed:
feat: implement candles data streaming and improve code quality

- Add candles data subscription and real-time streaming in ServiceHyperliquidSubscription
- Implement useCandles hook for candle data management with interval switching
- Add PerpCandles component for OHLCV data visualization and price tracking
- Enhance type definitions with proper interface naming conventions
- Fix extensive ESLint errors and apply Prettier formatting across codebase
- Refactor component structure: replace PerpChartsSection with PerpCandles

Minnzen committed:
feat: implement candles data streaming and improve code quality

- Add candles data subscription and real-time streaming in ServiceHyperliquidSubscription
- Implement useCandles hook for candle data management with interval switching
- Add PerpCandles component for OHLCV data visualization and price tracking
- Enhance type definitions with proper interface naming conventions
- Fix extensive ESLint errors and apply Prettier formatting across codebase
- Refactor component structure: replace PerpChartsSection with PerpCandles

ezailWang committed:
fix: perp order info panel coding (#8237)

Signed-off-by: ezailWang <jelly@onekey.so>

ezailWang committed:
fix: perp order info panel coding (#8237)

Signed-off-by: ezailWang <jelly@onekey.so>

ezailWang committed:
fix: perp tradesHistory list (#8248)

ezailWang committed:
fix: perp tradesHistory list (#8248)

ezailWang committed:
Perp orderinfo listopt (#8258)

ezailWang committed:
Perp orderinfo listopt (#8258)

huhuanming committed:
style: perp layout (#8257)

* fix: fix scrollview

* Update PerpAccountPanel.tsx

* Update PerpDesktopLayout.tsx

* Update PerpDesktopLayout.tsx

* Update HeaderRight.tsx

* Update PerpMobileLayout.tsx

* Update PerpMobileLayout.tsx

* Update PerpDesktopLayout.tsx

* Update PerpTickerBar.tsx

* Update CommonTableListView.tsx

huhuanming committed:
style: perp layout (#8257)

* fix: fix scrollview

* Update PerpAccountPanel.tsx

* Update PerpDesktopLayout.tsx

* Update PerpDesktopLayout.tsx

* Update HeaderRight.tsx

* Update PerpMobileLayout.tsx

* Update PerpMobileLayout.tsx

* Update PerpDesktopLayout.tsx

* Update PerpTickerBar.tsx

* Update CommonTableListView.tsx

Minnzen committed:
Merge branch 'perp' of https://github.com/OneKeyHQ/app-monorepo into perp

ezailWang committed:
fix: perps md (#8269)

Signed-off-by: ezailWang <jelly@onekey.so>

originalix committed:
Merge branch 'x' of github.com:OneKeyHQ/app-monorepo into perp

# Conflicts:
#	development/spellCheckerSkipWords.js

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquid.ts:
```diff
@@ -0,0 +1,145 @@
+  approveBuilderFee?: boolean;
+}
+
+const FALLBACK_BUILDER_ADDRESS = '0x9b12E858dA780a96876E3018780CF0D83359b0bb';
```
这是什么地址

Minnzen replied:
这个是我们的 builder code ，正常是从 DB 中获取，考虑边缘 case 添加了 fallback 地址以增强稳定性

originalix replied:
放到 `packages/shared/types/hyperliquid/perp.constants.ts` 吧

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquid.ts:
```diff
@@ -0,0 +1,145 @@
+      })
+      .catch((error) => {
+        console.error('Failed to load perp config:', error);
+      });
```
写一个 init 函数，继续用 async/await 的形式调用 simpleDb api 

在构造函数里：`void this.init()`

构造函数里写异步逻辑有点反模式

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquid.ts:
```diff
@@ -0,0 +1,145 @@
+      });
+  }
+
+  private get infoService(): ServiceHyperliquidInfo {
+    return this.backgroundApi.serviceHyperliquidInfo;
+  }
+
+  private get exchangeService(): ServiceHyperliquidExchange {
+    return this.backgroundApi.serviceHyperliquidExchange;
+  }
+
+  private get walletService(): ServiceHyperliquidWallet {
+    return this.backgroundApi.serviceHyperliquidWallet;
+  }
```
如果是这么使用的话，那建议 ServiceHyperliquidExchange 等几个剩余的 service 都写成 class，构造函数里也传入 backgroundApi

调用时的写法 `serviceHyperliquid.exchangeService.xxxxxxx`

这样 App 的业务 Service 只有 ServicePerps 一个，比较清晰

否则注入多个 Service，整体感觉有点散乱，看业务逻辑不清晰 @Minnzen

originalix replied:
下个 PR 重构多 service 的情况 @Minnzen ，我先 resolve 了

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquid.ts:
```diff
@@ -0,0 +1,145 @@
+    try {
+      const [extraAgents, maxBuilderFee] = await Promise.all([
+        this.infoService.getExtraAgents({
+          user: params.userAddress as `0x${string}`,
```
用 IHex 类型

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquid.ts:
```diff
@@ -0,0 +1,145 @@
+        tasks.push(
+          this.exchangeService.approveBuilderFee({
+            builder: this.builderAddress as `0x${string}`,
+            maxFeeRate: `${this.maxBuilderFee / 100}%`, // maxBuilderFee=40, maxFeeRate=0.04%
```
这地方期望传的是字符串的百分比么？

Minnzen replied:
是的，应该需要改成百分比数值保持一致性？

Minnzen replied:
这里有个情况是 HL 支持百分比 string 以及 int 的方式，但我们使用的 sdk 使用的是百分比 string ，且用了 @zuozhuo 的 getPerpConfig 的返回类型是 int ，因此做了层转换

Minnzen replied:
这里有个情况是 HL 支持百分比 string 以及 int 的方式，但我们使用的 sdk 使用的是百分比 string ，且用了 @sidmorizon 的 getPerpConfig 的返回类型是 int ，因此做了层转换

originalix replied:
sdk 使用 string 百分比那类型不用改，但是计算方式用 BigNumber 处理

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidExchange.ts:
```diff
@@ -0,0 +1,485 @@
+interface IAgentApprovalRequest {
+  agent: `0x${string}`;
+  authorize: boolean;
+}
```
类型定义都放 `shared/types/perps/*` 里， 另外 `0x${string}` 单独写一个类型: IHexAddress 这样，全局复用

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidSubscription.ts:
```diff
@@ -0,0 +1,526 @@
+      state.currentSymbol = params.currentSymbol;
+    }
+    if (params.currentCandleInterval !== undefined) {
+      state.currentCandleInterval = params.currentCandleInterval;
```
candle 相关逻辑可以去除，请求不发生在客户端，直接从 iframe 请求

Minnzen replied:
所有逻辑吗，包括 ui ? 或许可以先 merge 下

Minnzen replied:
移除了

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidSubscription.ts:
```diff
@@ -0,0 +1,526 @@
+  @backgroundMethod()
+  async reconnect(): Promise<void> {
+    await this.disconnect();
+    await new Promise((resolve) => setTimeout(resolve, 1000));
```
timerUtils.wait

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidWallet.ts:
```diff
@@ -0,0 +1,232 @@
+
+import type { IBackgroundApi } from '../../apis/IBackgroundApi';
+
+const CHAIN_ID = 'evm--42161'; // Arbitrum hex chainId
```
getNetworkIdsMap().arbitrum

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidWallet.ts:
```diff
@@ -0,0 +1,232 @@
+      message: value,
+    };
+
+    const address = await this.getAddress();
```
得考虑账户没地址的场景，可以先 throw error，后面业务逻辑开发完，补充账户相关逻辑

举例传入的 accountId 是一个 SUI 私钥账户的 id，那就获取不到 evm 地址， 或者传入的签名地址是 sui 的

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidSubscription.ts:
```diff
@@ -0,0 +1,526 @@
+
+  private _connectionMonitor: ReturnType<typeof setInterval> | null = null;
+
+  private _pendingUpdate: Promise<void> | null = null;
```
_pendingUpdate 有什么实际作用？没看到有调用关系

Minnzen replied:
这个 _pendingUpdate 是早期为了防止 updateSubscriptions 方法并发调用导致订阅状态混乱而设计的互斥锁。后期做了优化，因此不再使用这种方式，这个是遗留的无效代码，我清理下

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidSubscription.ts:
```diff
@@ -0,0 +1,526 @@
+  ): Promise<Promise<void>[]> {
+    if (toSubscribe.length === 0) return [];
+
+    const subscribePromises = toSubscribe.map(async (spec) => {
+      await this._createSubscription(spec);
+    });
+
+    return subscribePromises;
```
_executeSubscriptions 函数的 subscribePromises 不会等待订阅成功，会立即返回，是期望的么？

Minnzen replied:
这个符合预期，不过应该不再需要 return 了，也是一些修改遗留的部分，我改下

originalix commented on packages/kit/src/states/jotai/contexts/hyperliquid/actions.ts:
```diff
@@ -0,0 +1,650 @@
+    set(l2BookAtom(), data);
+  });
+
+  updateCandles = contextAtomMethod(
```
candles 相关函数移除，文件能减少 200 行代码

originalix commented on packages/kit/src/states/jotai/contexts/hyperliquid/atoms.ts:
```diff
@@ -0,0 +1,378 @@
+      return null;
+    }
+
+    const ctx = activeAssetCtx?.ctx;
+    const markPrice = ctx?.markPx || '0';
+    const oraclePrice = ctx?.oraclePx || '0';
+    const funding = ctx?.funding || '0';
+    const openInterest = ctx?.openInterest || '0';
+    const volume24h = ctx?.dayNtlVlm || '0';
+    const prevDayPx = ctx?.prevDayPx || markPrice;
+
+    const markPriceNum = parseFloat(markPrice);
+    const prevDayPxNum = parseFloat(prevDayPx);
+    const change24hPercent =
+      prevDayPxNum > 0
+        ? ((markPriceNum - prevDayPxNum) / prevDayPxNum) * 100
+        : 0;
+
+    const tradingData = {
+      ...tokenInfo,
+      ...activeAssetData,
+      name: currentToken,
+      markPx: markPrice,
+      markPrice,
+      oraclePrice,
+      prevDayPrice: prevDayPx,
+      fundingRate: funding,
+      openInterest,
+      volume24h,
+      change24hPercent,
+      lastUpdate: Date.now(),
+    };
```
跟上面逻辑一致，写一个子函数处理吧

originalix commented on packages/kit/src/views/Perp/pages/Perp.tsx:
```diff
@@ -0,0 +1,59 @@
+  return (
+    <Page>
+      <TabPageHeader
+        sceneName={EAccountSelectorSceneName.perp}
```
Perp 账户跟首页同步， 不用单独定义 `EAccountSelectorSceneName.perp`

使用 `EAccountSelectorSceneName.home`，让账户变动跟随首页

originalix commented on packages/kit/src/views/PerpTrade/pages/PageWebviewPerpTrade.tsx:
```diff
@@ -256,7 +256,7 @@ function WebviewPerpTradeView() {
     <Page fullPage>
       <TabPageHeader
-        sceneName={EAccountSelectorSceneName.home}
+        sceneName={EAccountSelectorSceneName.perp}
```
同上，这里不能改

Minnzen replied:
这个因为看到 webview 方案是这么使用的，或许那边也需要改下？

originalix commented on packages/kit/src/views/PerpTrade/pages/PageWebviewPerpTrade.tsx:
```diff
@@ -309,7 +309,7 @@ const PageWebviewPerpTrade = () => {
     <AccountSelectorProviderMirror
       config={{
-        sceneName: EAccountSelectorSceneName.home,
+        sceneName: EAccountSelectorSceneName.perp,
```
`EAccountSelectorSceneName.home`

ezailWang committed:
fix: perp trade history list modal (#8278)

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidExchange.ts:
```diff
@@ -0,0 +1,485 @@
+    slippage: number,
+  ): string {
+    const price = parseFloat(markPrice);
+    const slippageMultiplier = isBuy ? 1 + slippage : 1 - slippage;
```
涉及到数学计算都用 bignumber

Minnzen replied:
好🫡 有打算 review 一遍全部都改掉的

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidExchange.ts:
```diff
@@ -0,0 +1,485 @@
+
+      if (
+        'market' in params.orderType &&
+        (!params.limitPx || params.limitPx === '0')
```
bignumber 判断金额

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidExchange.ts:
```diff
@@ -0,0 +1,485 @@
+        t:
+          'limit' in params.orderType
+            ? {
+                limit: { tif: params.orderType.limit.tif as ITIF },
```
原类型定义不正确吗，这里必须要 as ？

Minnzen replied:
后续的代码已经修改且避免了 request params 的类型问题

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidExchange.ts:
```diff
@@ -0,0 +1,485 @@
+            trigger: {
+              isMarket,
+              triggerPx: params.tpTriggerPx,
+              tpsl: 'tp' as const,
```
修改原类型，避免 as

Minnzen replied:
有些其他地方的 as 也一并检查并处理了

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidExchange.ts:
```diff
@@ -0,0 +1,485 @@
+  ): Promise<IOrderResponse> {
+    this._ensureSetup();
+    const midPx = params.midPx;
+    const price = this._calculateSlippagePrice(
```
方法用对象参数，避免平铺

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidExchange.ts:
```diff
@@ -0,0 +1,485 @@
+      return await this.updateLeverage({
+        asset: params.assetId,
+        leverage: params.leverage,
+        isCross: params.isCross ?? true,
```
默认为 true 会反直觉，上层调用者以为不传是 false ，建议参数必须，不传就 报错

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidInfo.ts:
```diff
@@ -0,0 +1,100 @@
+      this._infoClient = new InfoClient({
+        transport,
+      }) as InfoClient;
+    }
```
为啥要 as

Minnzen replied:
最早在使用 sdk 的过程中对 sdk 的类型进行了一次包装，出现类型错误然后赶 demo 的临时解决方案，后续还没进行清理，现在已经处理了

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidSubscription.ts:
```diff
@@ -0,0 +1,526 @@
+    let sdkSubscription: any;
+
+    try {
+      switch (spec.type) {
```
用枚举

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidSubscription.ts:
```diff
@@ -0,0 +1,526 @@
+
+    this._connectionMonitor = setInterval(() => {
+      this._emitConnectionStatus();
+    }, 30_000);
```
定时器性能不好，用 global atom 更新 ui

sidmorizon commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidWallet.ts:
```diff
@@ -0,0 +1,232 @@
+    private accountId: string,
+    private backgroundApi: IBackgroundApi,
+  ) {
+    this.instanceId = `onekey-${Date.now()}-${Math.random()
```
random 用 uuid

sidmorizon commented on packages/shared/src/utils/tokenUtils.ts:
```diff
@@ -551,3 +551,17 @@ export function getTokenPriceChangeStyle({
+  if (!price || Number.isNaN(price)) return '0';
+  const precision = price.toPrecision(maxDigits);
+  const num = Number(precision);
+  let result = num.toString();
```
bignumber

sidmorizon commented on packages/kit/src/views/Perp/hooks/usePerpSession.ts:
```diff
@@ -0,0 +1,88 @@
+  }, [connectionState, currentUser]);
+
+  // Simplified actions for demo - in real app would use proper actions
+  const connect = async () => {
```
useCallback

Minnzen committed:
refactor: major code simplification and type safety improvements

- Introduce ESubscriptionType enum and centralize subscription constants
- Simplify service layer by consolidating redundant logic (net -400+ lines)
- Remove complex candles processing from actions, streamline data flow
- Replace string-based types with proper TypeScript interfaces (I-prefixed)
- Add perp.constants.ts for centralized configuration management
- Clean up PerpCandles component by removing redundant implementations
- Enhance type safety across SDK, service, and state management layers
- Consolidate subscription configuration with improved priority handling

originalix commented on packages/kit-bg/src/services/ServiceHyperLiquid/ServiceHyperliquidSubscription.ts:
```diff
@@ -0,0 +1,414 @@
+    return createSubscription(spec.type, client, spec.params, handleData);
+  }
+
+  private async _createSubscription(spec: ISubscriptionSpec): Promise<void> {
```
感觉现在想增加一个 websocket 的订阅，得改动好多地方的代码，例如需要监听实时的交易历史 `userFills` @Minnzen

Minnzen replied:
是，如果改动频率不大，或许是可以接受的？总共的订阅数量并不多，我看优化空间不大，还是有建议简化的地方么

Kahnchan committed:
refactor: enhance PerpTokenSelector component (#8285)

- Replace custom AssetIcon with Token component for improved token representation.
- Update layout and styling for better responsiveness and accessibility.
- Refactor search input to use SearchBar for a more consistent UI.
- Optimize token display with additional information such as funding rate and volume.
- Improve code readability and maintainability by removing unused imports and simplifying structures.

Kahnchan committed:
Fix/perp token selector (#8288)

* refactor: enhance PerpTokenSelector component

- Replace custom AssetIcon with Token component for improved token representation.
- Update layout and styling for better responsiveness and accessibility.
- Refactor search input to use SearchBar for a more consistent UI.
- Optimize token display with additional information such as funding rate and volume.
- Improve code readability and maintainability by removing unused imports and simplifying structures.

* refactor: enhance PerpTickerBar and PerpTokenSelector components

- Introduce NumberSizeableText for improved number formatting in PerpTickerBar.
- Update layout and styling for better responsiveness and accessibility in both components.
- Replace SizableText with appropriate sizes and formatting for consistency.
- Optimize token display in PerpTokenSelector with formatted volume and open interest.
- Clean up unused imports and streamline component structures for better maintainability.

Minnzen committed:
refactor: enhance type safety and code quality across infrastructure

- Add comprehensive TypeScript type declarations for Hyperliquid services
- Improve numerical precision by using BigNumber for price calculations
- Reorganize import statements and enhance code structure consistency
- Add missing event payload type definitions for better type safety
- Fix webview perp trade page scene configuration
- Enhance service constructor patterns with proper type annotations
- Consolidate code organization across API and service layers

originalix commented:
@coderabbitai resolve

originalix committed:
Merge branch 'x' of github.com:OneKeyHQ/app-monorepo into perp

# Conflicts:
#	packages/shared/src/eventBus/appEventBus.ts
#	packages/shared/src/eventBus/appEventBusNames.ts
#	packages/shared/src/utils/tokenUtils.ts

sidmorizon committed:
Merge branch 'x' into perp
